src/simularium/VisDataParse.ts

```diff
@@ -1,119 +1,108 @@
-                );
-            }
-
-            const agentSubSetArray = visData.splice(0, chunkLength); // cut off the array of 1 agent data from front of the array;
-            if (agentSubSetArray.length < AGENT_OBJECT_KEYS.length) {
-                const attemptedMapping = AGENT_OBJECT_KEYS.map(
-                    (name, index) => `${name}: ${agentSubSetArray[index]}<br />`
-                );
-                // will be caught by controller.changeFile(...).catch()
-                throw new FrontEndError(
-                    "Your data is malformed, there are less entries than expected for this agent. ",
-                    ErrorLevel.ERROR,
-                    `Example attempt to parse your data: <pre>${attemptedMapping.join(
-                        ""
-                    )}</pre>`
-                );
-            }
-
-            const agent = parseOneAgent(agentSubSetArray);
-            parsedAgentData.push(agent);
         }
 
-        const frameData: FrameData = {
-            time: frame.time,
-            frameNumber: frame.frameNumber,
-        };
+        agentCount++;
 
-        parsedAgentDataArray.push(parsedAgentData);
-        frameDataArray.push(frameData);
-    });
+        // Copy agent data
+        for (let i = 0; i < FRAME_DATA_SIZE; i++) {
+            view[writeIndex++] = frame.data[readIndex++];
+        }
 
-    return {
-        parsedAgentDataArray,
-        frameDataArray,
+        // Validate data integrity
+        if (--readIndex + nSubPoints > frame.data.length) {
+            throw new FrontEndError(
+                `Your data is malformed, there are too few entries. Found ${
+                    frame.data.length
+                }, expected ${readIndex + nSubPoints}.`,
+                ErrorLevel.ERROR
+            );
+        }
+
+        // Copy subpoints
+        for (let i = 0; i < nSubPoints; i++) {
+            view[writeIndex++] = frame.data[readIndex++];
+        }
+
+        // Adjust offsets relative to next agent's # of subpoints
+        offset += chunkLength;
+        nSubPoints = visData[offset + AGENT_OBJECT_KEYS.indexOf("nSubPoints")];
+        chunkLength = FRAME_DATA_SIZE + nSubPoints;
+        currentAgentData = visData.slice(offset, offset + chunkLength);
+    }
+
+    // Write header data
+    view[0] = frame.frameNumber;
+    view[1] = frame.time;
+    view[2] = agentCount;
+
+    const arrayBuffer: ArrayBuffer = view.buffer;
+    const frameData: CachedFrame = {
+        data: arrayBuffer,
+        frameNumber: frame.frameNumber,
+        time: frame.time,
+        agentCount: agentCount,
+        size: totalSize,
     };
+
+    return frameData;
+}
+
+function calculateBufferSize(data: number[]): number {
+    let size = AGENT_HEADER_SIZE * 4; // Header size in bytes
+    let index = 0;
+
+    while (index < data.length) {
+        size += FRAME_DATA_SIZE * 4; // Agent header size in bytes
+        const nSubPoints = data[index + FRAME_DATA_SIZE - 1];
```

Comment:
Nit: I noticed two different ways you compute the offset to subpoints.  Above here you also use `AGENT_OBJECT_KEYS.indexOf("nSubPoints")]` --  is that the same as `FRAME_DATA_SIZE-1` ?  Or are they doing different things in each place where they are used?

Reply:
Yeah they are the same, I'll change this one to match the others